src/Http/Paginators/OffsetRequestPaginator.php

```diff
@@ -0,0 +1,294 @@
+     */
+    public function totalPages(): int
+    {
+        // Make sure we have a response.
+        if (is_null($this->currentResponse)) {
+            $this->current();
+        }
+
+        return $this->lastPage();
+    }
+
+    /**
+     * @return int
+     */
+    public function firstPage(): int
+    {
+        // TODO: Or should we use something from the response?
+        return 1;
+    }
+
+    /**
+     * @return int
+     */
+    public function firstOffset(): int
+    {
+        // TODO: Or should we use something from the response?
+        return 0;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function previousPage(): ?int
+    {
+        $page = $this->currentPage();
+
+        return $page > $this->firstPage()
+            ? $page - 1
+            : null;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function previousOffset(): ?int
+    {
+        $previousOffset = $this->currentOffset() - $this->limit();
+
+        return $previousOffset > $this->firstOffset()
+            ? $previousOffset
+            : null;
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasPreviousOffset(): bool
+    {
+        return ! is_null($this->previousOffset());
+    }
+
+    /**
+     * @return int
+     */
+    public function currentPage(): int
+    {
+        return (int) ceil($this->totalEntries() / $this->totalPages());
+    }
+
+    /**
+     * @return int
+     */
+    public function currentOffset(): int
+    {
+        return $this->offset;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function nextPage(): ?int
+    {
+        $page = $this->currentPage();
+
+        return $page < $this->lastPage()
+            ? $page + 1
+            : null;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function nextOffset(): ?int
+    {
+        $nextOffset = $this->currentOffset() + $this->limit();
+
+        return $nextOffset < $this->lastOffset()
+            ? $nextOffset
+            : null;
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasNextOffset(): bool
+    {
+        return ! is_null($this->nextOffset());
+    }
+
+    /**
+     * @return int
+     */
+    public function lastPage(): int
+    {
+        return (int) floor($this->totalEntries() / $this->totalPages());
+    }
+
+    /**
+     * @return int
+     */
+    public function lastOffset(): int
+    {
+        return $this->totalEntries() - $this->limit();
+    }
+
+    protected function applyPaging(Request $request): void
+    {
+        $request->query()->merge([
+            'limit', $this->limit(),
+            'offset', $this->currentOffset(),
+        ]);
+    }
+
+    /**
+     * @return void
+     */
+    public function rewind(): void
+    {
+        if (! $this->shouldRewind()) {
+            return;
+        }
+
+        parent::rewind();
+
+        // TODO: Rewind completely, or rewind to originalOffset?
+        $this->offset = 0;
```

Comment:
I would say original offset

Reply:
Yeah, that's what I'm thinking.
If we were to continue the iteration somewhere, and original offset is not 0/the lowest offset.
_Rewinding_ should be to rewind to the start, imo.
But it could also give unpredictable behaviour, if the first position (original offset/page) is not the lowest, then we'd still send request to the lowest..